//definimos el paquete para evitar conflictos de nombres
syntax = "proto3"; //definimos la version de proto a utilizar
package proto_services;

import "google/protobuf/empty.proto"; // para definir un mensaje vacio

option java_multiple_files = true; // genera un archivo por cada mensaje
option java_package = "com.example.proto.services"; //paquete de la clase generada
option java_outer_classname = "ServicesProto"; //nombre de la clase wrapper generada


service ChatService {
    rpc Chat(stream ChatMessage) returns (stream ChatMessage); //metodo stream que recibe y devuelve mensajes de chat
    rpc ConnectToRoom(ConnectToRoomRequest) returns (ConnectToRoomResponse); //metodo que recibe una peticion de conexion a una sala y devuelve una respuesta de conexion
    rpc GetRooms(google.protobuf.Empty) returns (GetRoomsResponse); //metodo que recibe devuelve las salas disponibles
}

service UserService {
    rpc RegisterUser(UserRegistration) returns (User); //metodo que recibe un usuario y devuelve el usuario registrado
    rpc LoginUser(UserLogin) returns (User); //metodo que recibe un usuario y devuelve el usuario logueado
}

message ChatMessage{
    User sender = 1; //usuario que envia el mensaje
    string message = 2; //mensaje
    Room room = 3; //sala a la que pertenece el mensaje
}

message Room{
    int32 id = 1; //id de la sala
    string name = 2; //nombre de la sala
    string description = 3; //descripcion de la sala
}
message GetRoomsResponse{
    repeated Room rooms = 1; //lista de salas
}

message ConnectToRoomRequest{
    int32 room = 1; //sala a la que nos queremos conectar
    User user = 2; //usuario que se conecta
}

message ConnectToRoomResponse{
    Room room = 1; //sala a la que nos conectamos
}

message UserRegistration{
    string username = 1; //nombre de usuario
    string email = 2; //email del usuario
    string password = 3; //contraseña del usuario
}

message UserLogin{
    string email = 1; //email del usuario
    string password = 2; //contraseña del usuario
}

message User {
    int32 id = 1; //id del usuario
    UserRegistration registration = 2; //informacion de registro del usuario
}